### PR TITLE
Byzantine Primary for PRA

### PR DESCRIPTION
**Pull request by arc-vmware**
PreExecution Result Authentication require a
Byzantine Primary Replica.
This feature should work even if the Byzantine
Primary Replica is present in the network.All
the apollo test suite goes into view change and
then re-elect a new primary due to this feature.
If this feature is off, the same Byzantine Replica
is creating a deferred consensus. Due to this
phenomenon the test will not stop as clients will
wait for the result for a long time.

A new strategy is created for this Byzantine
Replica.

### CODE REVIEW

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+LONG_REQ_TIMEOUT_MILLI = 15000
+ASSERT_TIMEOUT_SEC = 90
+
+def start_replica_cmd(builddir, replica_id, view_change_timeout_milli="10000"):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+    The replica is started with a short view change timeout.
+    Note each arguments is an element in a list.
+    """
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_byzantine_primary_preexecution.py:38]: can you please document what is the byzantine behavior if this function is used for starting replicas?

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                    err_msg="Make sure view change has occurred."
+                )
+                new_primary = await bft_network.get_current_primary()
+                await bft_network.init_preexec_count(new_primary)
+                await trio.sleep(seconds=viewchange_timeout_secs)
+                return new_primary
+        except trio.TooSlowError:
+            return self.check_viewchange_noexcept(bft_network, initial_primary, viewchange_timeout_secs)
+
+    async def send_single_read(self, skvbc, client):
```
Comment (tdimitrov) [inline tests/apollo/test_skvbc_byzantine_primary_preexecution.py:99]: I see that send_single_read, send_single_write_with_pre_execution_and_kv, send_single_write_with_pre_execution, etc are 1:1 with the ones from test_skvbc_preexecution.py. Can you move them in separate module?

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+    return cmd
+
+def start_replica_cmd_assymetric_communication(builddir, replica_id, view_change_timeout_milli="10000"):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+    The replica is started with a short view change timeout.
+    Note each arguments is an element in a list.
+    """
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_byzantine_primary_preexecution.py:61]: same here - document the behavior

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
+# Concord
```
Comment (tdimitrov) [inline tests/apollo/test_skvbc_byzantine_primary_preexecution.py:1]: I see a lot of shared code between test_skvbc_batch_preexecution.py, test_skvbc_byzantine_primary_preexecution.py and test_skvbc_preexecution.py.
We have to figure out a way to reuse the code better. One more test suite and it will be unmaintainable.

This is out of scope for this PR of course.

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+                        bft_network.all_replicas(without={initial_primary})),
+                    expected=lambda v: v != initial_primary,
+                    err_msg="Make sure view change has occurred."
+                )
+                new_primary = await bft_network.get_current_primary()
+                await bft_network.init_preexec_count(new_primary)
+                await trio.sleep(seconds=viewchange_timeout_secs)
+                return new_primary
+        except trio.TooSlowError:
+            return self.check_viewchange_noexcept(bft_network, initial_primary, viewchange_timeout_secs)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_byzantine_primary_preexecution.py:97]: are you using recursion here?

If so, note that `self.check_viewchange_noexcept()` is async, so you need to call it with `await`. Otherwise it won't be executed ...

**Review by teoparvanov — COMMENTED** [review]
Amitava, this is a good start... However, I'd suggest to have a smaller, more focused set of tests (2-3), which test the feature that Tsvetomir developed.

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py (hunk 1 above)
Comment (arc-vmware) [reply tests/apollo/test_skvbc_byzantine_primary_preexecution.py:38]: I will document the Byzantine behaviour for all test cases. After you pointed out, I found its missing for all the functions. Thanks @teoparvanov

**Review by tdimitrov — APPROVED** [review]
Good job @arc-vmware . I'm approving but please address the request for the copy-pasted functions.

File: tests/simpleKVBC/TesterReplica/strategy/ShufflePreProcessMsgStrategy.hpp [hunk 6]
```diff
@@ -11,17 +11,20 @@
...
 
 namespace concord::kvbc::strategy {
 class ShufflePrePrepareMsgStrategy : public IByzantineStrategy {
  public:
   std::string getStrategyName() override;
   uint16_t getMessageCode() override;
-  bool changeMessage(std::shared_ptr<MessageBase> &msg) override;
-  ShufflePrePrepareMsgStrategy() = default;
+  bool changeMessage(std::shared_ptr<bftEngine::impl::MessageBase> &msg) override;
+  explicit ShufflePrePrepareMsgStrategy(logging::Logger &logger) : logger_(logger) {}
```
Comment (teoparvanov) [inline tests/simpleKVBC/TesterReplica/strategy/ShufflePreProcessMsgStrategy.hpp:23]: please rename the .hpp and .cpp files accordingly : ShufflePrePrepareMsgStrategy.hpp / ShufflePrePrepareMsgStrategy.cpp

File: tests/simpleKVBC/TesterReplica/strategy/MangledPreProcessResultMsgStrategy.hpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#pragma once
+
+#include "Logger.hpp"
+#include "TesterReplica/strategy/ByzantineStrategy.hpp"
+
+namespace concord::kvbc::strategy {
+class MangledPreProcessResultMsgStrategy : public IByzantineStrategy {
```
Comment (teoparvanov) [inline tests/simpleKVBC/TesterReplica/strategy/MangledPreProcessResultMsgStrategy.hpp:19]: As discussed over slack, please add documentation strings for both strategies (shuffle/mangle) in the header files.

File: tests/apollo/util/bft.py [hunk 8]
```diff
@@ -1341,13 +1341,15 @@ async def check_initital_key_exchange(self, stop_replicas=True):
...
             total_pre_exec_requests_executed = 0
-            with trio.fail_after(10):
-                while pre_proc_req < num_requests or \
-                        total_pre_exec_requests_executed < num_requests:
+            threshold_request = num_requests * threshold;
+            with trio.fail_after(timeout_in_sec):
+                while pre_proc_req < threshold_request or \
+                        total_pre_exec_requests_executed < threshold_request:
                     key1 = ["preProcessor", "Counters", "preProcReqCompleted"]
                     pre_proc_req = await self.metrics.get(replica_id, *key1) - self.initial_preexec_sent
```
Comment (teoparvanov) [inline tests/apollo/util/bft.py:1354]: if possible, please revert the change in the shared utilities code. If you really need this change, I'd suggest to have it carefully reviewed by @KTeofanidis ...

File: tests/simpleKVBC/TesterReplica/WrapCommunication.hpp [hunk 9]
```diff
@@ -72,7 +58,9 @@ class WrapCommunication : public ICommunication {
 
  private:
   std::unique_ptr<bft::communication::ICommunication> communication_;
+  bool separate_communication_;
```
[the remark this answers is not in the document]
Comment (arc-vmware) [reply tests/simpleKVBC/TesterReplica/WrapCommunication.hpp:61]: Yes, I will add the comment, thanks @tdimitrov

File: tests/simpleKVBC/TesterReplica/setup.cpp [hunk 10]
```diff
@@ -106,7 +111,7 @@ std::unique_ptr<TestSetup> TestSetup::ParseArgs(int argc, char** argv) {
     int o = 0;
     int optionIndex = 0;
     LOG_INFO(GL, "Command line options:");
-    while ((o = getopt_long(argc, argv, "i:k:n:s:v:a:3:l:e:w:c:b:m:q:z:y:up:t:o:r:g:x", longOptions, &optionIndex)) !=
+    while ((o = getopt_long(argc, argv, "i:k:n:s:v:a:3:l:e:w:c:b:m:q:z:y:udp:t:o:r:g:x", longOptions, &optionIndex)) !=
```
[the remark this answers is not in the document]
Comment (arc-vmware) [reply tests/simpleKVBC/TesterReplica/setup.cpp:114]: We will use capital letters later. 
Our replica config is becoming like a novel, but this is common for any big system :)

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py (hunk 4 above)
Comment (arc-vmware) [reply tests/apollo/test_skvbc_byzantine_primary_preexecution.py:1]: This is in this scope. Thanks for the comment. I will actually remove all the testcases and keep only first two as discussed with @teoparvanov . This will simplify the PR, we will be able to test the feature going for 1.4. Reduce the time to complete the build. Later this feature can be used by other platforms of testing who are using skvbc_replica. They can ask for any strategy and add Byzantine threats and do more testing. This PR and its sister PR is to create a boilerplate to add this capability. Thanks @tdimitrov for raising this.

File: tests/apollo/util/bft.py (hunk 8 above)
Comment (arc-vmware) [reply tests/apollo/util/bft.py:1354]: This change will not be needed as we are reducing the number of tests.

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py (hunk 5 above)
Comment (arc-vmware) [reply tests/apollo/test_skvbc_byzantine_primary_preexecution.py:97]: Thanks @teoparvanov for this comment, good catch.

File: tests/apollo/test_skvbc_byzantine_primary_preexecution.py (hunk 2 above)
Comment (arc-vmware) [reply tests/apollo/test_skvbc_byzantine_primary_preexecution.py:99]: I will remove those functions as for testing the specific functionality, these functions are not needed.

File: tests/simpleKVBC/TesterReplica/strategy/ShufflePreProcessMsgStrategy.hpp (hunk 6 above)
Comment (arc-vmware) [reply tests/simpleKVBC/TesterReplica/strategy/ShufflePreProcessMsgStrategy.hpp:23]: Ok, I will rename the files.

File: tests/simpleKVBC/TesterReplica/strategy/MangledPreProcessResultMsgStrategy.hpp (hunk 7 above)
Comment (arc-vmware) [reply tests/simpleKVBC/TesterReplica/strategy/MangledPreProcessResultMsgStrategy.hpp:19]: I have added documentation for both. I have updated the strategy in the changeMessage method and added a general comment on Shuffle header.

**Review by teoparvanov — APPROVED** [review]
Looks great, thank you @arc-vmware !
